pkg/reconciler/v1alpha1/revision/reconcile_resources.go

```diff
@@ -119,6 +81,66 @@ func (c *Reconciler) reconcileDeployment(ctx context.Context, rev *v1alpha1.Revi
 	return nil
 }
 
+func (c *Reconciler) deployment(ctx context.Context, rev *v1alpha1.Revision, deploymentName string, logger *zap.SugaredLogger) (*v1.Deployment, error) {
+	ns := rev.Namespace
+	deployment, err := c.deploymentLister.Deployments(ns).Get(deploymentName)
+	if apierrs.IsNotFound(err) {
+		// Deployment does not exist. Create it.
+		rev.Status.MarkDeploying("Deploying")
+		deployment, err = c.createDeployment(ctx, rev)
+		if err != nil {
+			logger.Errorf("Error creating deployment %q: %v", deploymentName, err)
+			return nil, err
+		}
+		logger.Infof("Created deployment %q", deploymentName)
+	} else if err != nil {
+		logger.Errorf("Error reconciling deployment %q: %v", deploymentName, err)
+		return nil, err
+	} else if !metav1.IsControlledBy(deployment, rev) {
+		// Surface an error in the revision's status, and return an error.
+		rev.Status.MarkResourceNotOwned("Deployment", deploymentName)
+		return nil, fmt.Errorf("Revision: %q does not own Deployment: %q", rev.Name, deploymentName)
+	} else {
+		// The deployment exists, but make sure that it has the shape that we expect.
+		deployment, err = c.checkAndUpdateDeployment(ctx, rev, deployment)
+		if err != nil {
+			logger.Errorf("Error updating deployment %q: %v", deploymentName, err)
+			return nil, err
+		}
+	}
+	return deployment, nil
+}
+
+func (c *Reconciler) handleOutOfSyncDeployment(rev *v1alpha1.Revision, deployment *v1.Deployment, logger *zap.SugaredLogger) {
+	if !replicasOutOfSync(deployment) {
+		return
+	}
+
+	// If a container keeps crashing (no active pods in the deployment although we want some)
+	ns := rev.Namespace
+	pods, err := c.KubeClientSet.CoreV1().Pods(ns).List(metav1.ListOptions{LabelSelector: metav1.FormatLabelSelector(deployment.Spec.Selector)})
+	if err != nil {
+		logger.Errorf("Error getting pods: %v", err)
+	} else if len(pods.Items) > 0 {
+		// Arbitrarily grab the very first pod, as they all should be crashing
+		markContainerCrashing(rev, pods.Items[0])
+	}
+}
+func markContainerCrashing(rev *v1alpha1.Revision, pod corev1.Pod) {
+	for _, status := range pod.Status.ContainerStatuses {
+		if status.Name == resources.UserContainerName {
+			if t := status.LastTerminationState.Terminated; t != nil {
+				rev.Status.MarkContainerExiting(t.ExitCode, t.Message)
+			}
+			return
```

Comment:
should you only return if you actually set the status?

Reply:
well, that doesn't mean it was correct in the first place :)

Reply:
Created a PR to address the logical change: http://go/gh/knative/serving/pull/3552

Reply:
I think this change is correct as-is. The check is only performed on the user-container and we can exit as early as the user-container was found, be it crashed or not.

However, the name of this function is misleading in that it doesn't always mark the container as crashing but only if the container is actually crashing.

Reply:
Updated the function name